File: src/系统/03．技能系统/00．技能模板+函数/01．技能函数/09．形状区域/胶囊区域.lua
local ____lualib = require("lualib_bundle")
local __TS__SparseArrayNew = ____lualib.__TS__SparseArrayNew
local __TS__SparseArrayPush = ____lualib.__TS__SparseArrayPush
local __TS__SparseArraySpread = ____lualib.__TS__SparseArraySpread
local ____exports = {}
--- 形状区域 - 胶囊形 / 线段宽度区域
-- 
-- 说明：
-- 1. 形状等价于“中间一段线 + 两端半圆”。
-- 2. 适合检测“沿路径扫过、但到结算时只看是否仍停留在路径宽度内”的目标。
-- 3. 先以线段中点粗筛，再做投影精判。
local jass = require("jass.common")
local ____require_result_0 = require("lib.扩展函数.自定义扩展函数.01．选取中心范围")
local getUnitsInRange = ____require_result_0.getUnitsInRange
local GetUnitX = jass.GetUnitX
local GetUnitY = jass.GetUnitY
local CreateGroup = jass.CreateGroup
local GroupAddUnit = jass.GroupAddUnit
local function _____8BA1_7B97_5E73_65B9_6839(_____503C)
    return jass.SquareRoot(_____503C)
end
local function _____8BA1_7B97_5750_6807_8DDD_79BB(x1, y1, x2, y2)
    local dx = x2 - x1
    local dy = y2 - y1
    return _____8BA1_7B97_5E73_65B9_6839(dx * dx + dy * dy)
end
local function _____5355_4F4D_662F_5426_5728_7EBF_6BB5_5BBD_5EA6_533A_57DF_5185_90E8(_____5355_4F4D, _____8D77_70B9X, _____8D77_70B9Y, _____7EC8_70B9X, _____7EC8_70B9Y, _____534A_5BBD, _____5305_542B_8FB9_754C)
    if _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return false
    end
    if _____534A_5BBD <= 0 then
        return false
    end
    local _____7EBF_6BB5X = _____7EC8_70B9X - _____8D77_70B9X
    local _____7EBF_6BB5Y = _____7EC8_70B9Y - _____8D77_70B9Y
    local _____7EBF_6BB5_957F_5EA6_5E73_65B9 = _____7EBF_6BB5X * _____7EBF_6BB5X + _____7EBF_6BB5Y * _____7EBF_6BB5Y
    if _____7EBF_6BB5_957F_5EA6_5E73_65B9 <= 0.0001 then
        local dx = GetUnitX(_____5355_4F4D) - _____8D77_70B9X
        local dy = GetUnitY(_____5355_4F4D) - _____8D77_70B9Y
        local _____8DDD_79BB_5E73_65B9 = dx * dx + dy * dy
        local _____534A_5BBD_5E73_65B9 = _____534A_5BBD * _____534A_5BBD
        local _____5305_542B_8FB9_754C_1
        if _____5305_542B_8FB9_754C then
            _____5305_542B_8FB9_754C_1 = _____8DDD_79BB_5E73_65B9 <= _____534A_5BBD_5E73_65B9
        else
            _____5305_542B_8FB9_754C_1 = _____8DDD_79BB_5E73_65B9 < _____534A_5BBD_5E73_65B9
        end
        return _____5305_542B_8FB9_754C_1
    end
    local _____70B9X = GetUnitX(_____5355_4F4D)
    local _____70B9Y = GetUnitY(_____5355_4F4D)
    local _____5230_8D77_70B9X = _____70B9X - _____8D77_70B9X
    local _____5230_8D77_70B9Y = _____70B9Y - _____8D77_70B9Y
    local _____6295_5F71_6BD4_4F8B = (_____5230_8D77_70B9X * _____7EBF_6BB5X + _____5230_8D77_70B9Y * _____7EBF_6BB5Y) / _____7EBF_6BB5_957F_5EA6_5E73_65B9
    if _____6295_5F71_6BD4_4F8B < 0 then
        _____6295_5F71_6BD4_4F8B = 0
    elseif _____6295_5F71_6BD4_4F8B > 1 then
        _____6295_5F71_6BD4_4F8B = 1
    end
    local _____6700_8FD1_70B9X = _____8D77_70B9X + _____7EBF_6BB5X * _____6295_5F71_6BD4_4F8B
    local _____6700_8FD1_70B9Y = _____8D77_70B9Y + _____7EBF_6BB5Y * _____6295_5F71_6BD4_4F8B
    local dx = _____70B9X - _____6700_8FD1_70B9X
    local dy = _____70B9Y - _____6700_8FD1_70B9Y
    local _____8DDD_79BB_5E73_65B9 = dx * dx + dy * dy
    local _____534A_5BBD_5E73_65B9 = _____534A_5BBD * _____534A_5BBD
    if _____5305_542B_8FB9_754C then
        return _____8DDD_79BB_5E73_65B9 <= _____534A_5BBD_5E73_65B9
    end
    return _____8DDD_79BB_5E73_65B9 < _____534A_5BBD_5E73_65B9
end
____exports["单位是否在胶囊区域"] = function(_____5355_4F4D, _____8D77_70B9X, _____8D77_70B9Y, _____7EC8_70B9X, _____7EC8_70B9Y, _____5BBD_5EA6, _____5305_542B_8FB9_754C)
    if _____5305_542B_8FB9_754C == nil then
        _____5305_542B_8FB9_754C = true
    end
    return _____5355_4F4D_662F_5426_5728_7EBF_6BB5_5BBD_5EA6_533A_57DF_5185_90E8(
        _____5355_4F4D,
        _____8D77_70B9X,
        _____8D77_70B9Y,
        _____7EC8_70B9X,
        _____7EC8_70B9Y,
        _____5BBD_5EA6 / 2,
        _____5305_542B_8FB9_754C
    )
end
____exports["获取胶囊区域单位"] = function(_____53C2_6570)
    if _____53C2_6570["宽度"] <= 0 then
        return {}
    end
    local _____4E2D_5FC3X = (_____53C2_6570["起点X"] + _____53C2_6570["终点X"]) / 2
    local _____4E2D_5FC3Y = (_____53C2_6570["起点Y"] + _____53C2_6570["终点Y"]) / 2
    local _____7EBF_6BB5_957F_5EA6 = _____8BA1_7B97_5750_6807_8DDD_79BB(_____53C2_6570["起点X"], _____53C2_6570["起点Y"], _____53C2_6570["终点X"], _____53C2_6570["终点Y"])
    local _____7C97_7B5B_534A_5F84 = _____7EBF_6BB5_957F_5EA6 / 2 + _____53C2_6570["宽度"] / 2
    local _____5019_9009_5355_4F4D = getUnitsInRange(_____4E2D_5FC3X, _____4E2D_5FC3Y, _____7C97_7B5B_534A_5F84)
    local _____7ED3_679C = {}
    for ____, _____5355_4F4D in ipairs(_____5019_9009_5355_4F4D) do
        do
            local ____exports__5355_4F4D_662F_5426_5728_80F6_56CA_533A_57DF_4 = ____exports["单位是否在胶囊区域"]
            local ____array_3 = __TS__SparseArrayNew(
                _____5355_4F4D,
                _____53C2_6570["起点X"],
                _____53C2_6570["起点Y"],
                _____53C2_6570["终点X"],
                _____53C2_6570["终点Y"],
                _____53C2_6570["宽度"]
            )
            local ____53C2_6570__5305_542B_8FB9_754C_2 = _____53C2_6570["包含边界"]
            if ____53C2_6570__5305_542B_8FB9_754C_2 == nil then
                ____53C2_6570__5305_542B_8FB9_754C_2 = true
            end
            __TS__SparseArrayPush(____array_3, ____53C2_6570__5305_542B_8FB9_754C_2)
            if not ____exports__5355_4F4D_662F_5426_5728_80F6_56CA_533A_57DF_4(__TS__SparseArraySpread(____array_3)) then
                goto __continue14
            end
            if _____53C2_6570["单位筛选"] ~= nil and not _____53C2_6570["单位筛选"](_____5355_4F4D) then
                goto __continue14
            end
            _____7ED3_679C[#_____7ED3_679C + 1] = _____5355_4F4D
        end
        ::__continue14::
    end
    return _____7ED3_679C
end
____exports["创建胶囊单位组"] = function(_____53C2_6570)
    local _____5355_4F4D_7EC4 = CreateGroup()
    local _____5355_4F4D_5217_8868 = ____exports["获取胶囊区域单位"](_____53C2_6570)
    for ____, _____5355_4F4D in ipairs(_____5355_4F4D_5217_8868) do
        GroupAddUnit(_____5355_4F4D_7EC4, _____5355_4F4D)
    end
    return _____5355_4F4D_7EC4
end
return ____exports

File: TS/系统/03．技能系统/00．技能模板+函数/01．技能函数/09．形状区域/胶囊区域.ts
/** @noSelfInFile */
/**
 * 形状区域 - 胶囊形 / 线段宽度区域
 *
 * 说明：
 * 1. 形状等价于“中间一段线 + 两端半圆”。
 * 2. 适合检测“沿路径扫过、但到结算时只看是否仍停留在路径宽度内”的目标。
 * 3. 先以线段中点粗筛，再做投影精判。
 */

const jass = require("jass.common") as any;
import type { 英雄技能距离修正上下文 } from "../../04．机制组件/11．技能属性修正";

const { getUnitsInRange } = require("lib.扩展函数.自定义扩展函数.01．选取中心范围") as {
  getUnitsInRange: (this: void, x: number, y: number, radius: number) => any[];
};
const { 按英雄技能距离修正上下文修正距离 } = require("系统.03．技能系统.00．技能模板+函数.04．机制组件.11．技能属性修正.index") as {
  按英雄技能距离修正上下文修正距离: (this: void, 基础距离: number, 上下文: any, 默认用途?: string) => number;
};

const GetUnitX = jass.GetUnitX as (u: any) => number;
const GetUnitY = jass.GetUnitY as (u: any) => number;
const CreateGroup = jass.CreateGroup as () => any;
const GroupAddUnit = jass.GroupAddUnit as (whichGroup: any, whichUnit: any) => void;

export interface 胶囊区域参数 {
  起点X: number;
  起点Y: number;
  终点X: number;
  终点Y: number;
  宽度: number;
  英雄技能距离修正?: 英雄技能距离修正上下文;
  单位筛选?: (this: void, 单位: any) => boolean;
  包含边界?: boolean;
}

function 计算平方根(值: number): number {
  return jass.SquareRoot(值) as number;
}

function 计算坐标距离(x1: number, y1: number, x2: number, y2: number): number {
  const dx = x2 - x1;
  const dy = y2 - y1;
  return 计算平方根(dx * dx + dy * dy);
}

function 单位是否在线段宽度区域内部(
  单位: any,
  起点X: number,
  起点Y: number,
  终点X: number,
  终点Y: number,
  半宽: number,
  包含边界: boolean
): boolean {
  if (单位 == null || 单位 === 0) return false;
  if (半宽 <= 0) return false;

  const 线段X = 终点X - 起点X;
  const 线段Y = 终点Y - 起点Y;
  const 线段长度平方 = 线段X * 线段X + 线段Y * 线段Y;
  if (线段长度平方 <= 0.0001) {
    const dx = GetUnitX(单位) - 起点X;
    const dy = GetUnitY(单位) - 起点Y;
    const 距离平方 = dx * dx + dy * dy;
    const 半宽平方 = 半宽 * 半宽;
    return 包含边界 ? 距离平方 <= 半宽平方 : 距离平方 < 半宽平方;
  }

  const 点X = GetUnitX(单位);
  const 点Y = GetUnitY(单位);
  const 到起点X = 点X - 起点X;
  const 到起点Y = 点Y - 起点Y;
  let 投影比例 = (到起点X * 线段X + 到起点Y * 线段Y) / 线段长度平方;
  if (投影比例 < 0) {
    投影比例 = 0;
  } else if (投影比例 > 1) {
    投影比例 = 1;
  }

  const 最近点X = 起点X + 线段X * 投影比例;
  const 最近点Y = 起点Y + 线段Y * 投影比例;
  const dx = 点X - 最近点X;
  const dy = 点Y - 最近点Y;
  const 距离平方 = dx * dx + dy * dy;
  const 半宽平方 = 半宽 * 半宽;

  if (包含边界) {
    return 距离平方 <= 半宽平方;
  }
  return 距离平方 < 半宽平方;
}

export function 单位是否在胶囊区域(
  单位: any,
  起点X: number,
  起点Y: number,
  终点X: number,
  终点Y: number,
  宽度: number,
  包含边界: boolean = true
): boolean {
  return 单位是否在线段宽度区域内部(
    单位,
    起点X,
    起点Y,
    终点X,
    终点Y,
    宽度 / 2,
    包含边界
  );
}

export function 获取胶囊区域单位(参数: 胶囊区域参数): any[] {
  if (参数.宽度 <= 0) {
    return [];
  }

  const 原线段长度 = 计算坐标距离(参数.起点X, 参数.起点Y, 参数.终点X, 参数.终点Y);
  const 线段长度 = 按英雄技能距离修正上下文修正距离(原线段长度, 参数.英雄技能距离修正, "胶囊长度");
  const 方向X = 原线段长度 > 0 ? (参数.终点X - 参数.起点X) / 原线段长度 : 0;
  const 方向Y = 原线段长度 > 0 ? (参数.终点Y - 参数.起点Y) / 原线段长度 : 0;
  const 终点X = 参数.起点X + 方向X * 线段长度;
  const 终点Y = 参数.起点Y + 方向Y * 线段长度;
  const 中心X = (参数.起点X + 终点X) / 2;
  const 中心Y = (参数.起点Y + 终点Y) / 2;
  const 粗筛半径 = 线段长度 / 2 + 参数.宽度 / 2;
  const 候选单位 = getUnitsInRange(中心X, 中心Y, 粗筛半径);
  const 结果: any[] = [];

  for (const 单位 of 候选单位) {
    if (!单位是否在胶囊区域(
      单位,
      参数.起点X,
      参数.起点Y,
      终点X,
      终点Y,
      参数.宽度,
      参数.包含边界 ?? true
    )) {
      continue;
    }

    if (参数.单位筛选 != null && !参数.单位筛选(单位)) {
      continue;
    }

    结果.push(单位);
  }

  return 结果;
}

export function 创建胶囊单位组(参数: 胶囊区域参数): any {
  const 单位组 = CreateGroup();
  const 单位列表 = 获取胶囊区域单位(参数);

  for (const 单位 of 单位列表) {
    GroupAddUnit(单位组, 单位);
  }

  return 单位组;
}

export {};
